File: src/0020_valid_parentheses.py
# Stack
def isValid(s: str) -> bool:
    hashmap = {
        ")": "(",
        "]": "[",
        "}": "{",
    }
    stack = []

    for c in s:
        if c in hashmap:
            if stack and stack[-1] == hashmap[c]:
                stack.pop()
            else:
                return False
        else:
            stack.append(c)

    return True if not stack else False


print(isValid("()"))  # True
print(isValid("()[]{}"))  # True
print(isValid("(]"))  # False

File: src/cpp/0020_valid_parentheses.cc
#include <string>
#include <stack>
#include <unordered_map>
#include <cassert>
using namespace std;

class Solution
{
public:
    bool isValid(string s)
    {
        unordered_map<char, char> map = {
            {')', '('},
            {']', '['},
            {'}', '{'}};

        stack<char> stack;

        for (char c : s)
        {
            if (map.find(c) != map.end())
            {
                char topElement = stack.empty() ? '#' : stack.top();
                if (topElement != map[c])
                {
                    return false;
                }
                stack.pop();
            }
            else
            {
                stack.push(c);
            }
        }

        return stack.empty();
    }
};

int main()
{
    Solution s;
    assert(s.isValid("()") == true);
    assert(s.isValid("()[]{}") == true);
    assert(s.isValid("(]") == false);
    assert(s.isValid("([)]") == false);
    assert(s.isValid("{[]}") == true);
    return 0;
}
